logging: syst: fix logging v2 with strings as arguments

dcpleung opened the pull request:
The packaged string coming from logging subsys requires some
additional processing if there are string arguments.
These strings are actually embedded inside the package so
we need to replace the string pointers inside the argument
list so they actually point to those strings. Without this
extra step of processing, MIPI Sys-T's printf function would
process these arguments directly which may be NULL pointers
or invalid one pointing to somewhere. This adds the processing
so that log messages with string arguments can be outputted
correctly through MIPI.

dcpleung committed:
samples: logging/syst: add more lines to show string processing

This adds a few log lines to show how string processing works
with Sys-T.

Note that CONFIG_LOG_BLOCK_IN_THREAD=y to make sure all log
messages are printed instead of being dropped.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
samples: logging/syst: expand to doing v1/v2 immediate/deferred

This expands the sample to do both v1 and v2, and both immediate
and deferred modes for each logging verion. This is to make sure
MIPI Sys-T output is not broken for all combinations. Default is
v2/immediate using prj.conf. Deferred mode is enabled via
overlay.

Also add qemu_x86 as an allowed platform as the samples are
working there too.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

nordic-krch commented on subsys/logging/log_output_syst.c:
```diff
@@ -786,27 +786,97 @@ void log_output_msg2_syst_process(const struct log_output *output,
+		 * [[0] * sizeof(uint32_t)]: Start of appended strings.
+		 */
+
+		uint8_t num_strings = data[1];
```
This algorithm is already present in cbprintf https://github.com/zephyrproject-rtos/zephyr/blob/fdc25cd44c2b693aecff96d3dfcd0ee088d48d5b/lib/os/cbprintf_packaged.c#L602
but it is a part of formatting function. Seems like it would be worth to extract it and make public. Not sure what would be a good name for it. Something like `cbprintf_locate` or `cbrprintf_settle`?

It could be called by logging core before passing message to the backends.

nordic-krch commented on samples/subsys/logging/syst/sample.yaml:
```diff
@@ -1,18 +1,36 @@
-  sample.logger.syst.v1:
-    platform_allow: mps2_an385 sam_e70_xplained
+  sample.logger.syst.v1.deferred:
+    extra_args: CONF_FILE=prj_v1_deferred.conf
```
why not use `OVERLAY_CONFIG` here and keep all common stuff in `prj.conf`. Now there is a lot of redundancy in those config files.

dcpleung replied:
OVERLAY! That's the keyword I was looking for. Thanks. Updated.

nordic-krch commented on subsys/logging/log_output_syst.c:
```diff
@@ -786,27 +786,97 @@ void log_output_msg2_syst_process(const struct log_output *output,
 
-		fmt = ((char **)buf)[1];
-		buf += sizeof(char *) * 2;
+		if (output_buf != NULL) {
```
There are corner cases for preparing va_list (xtensa, x86_64), see set of `cbprintf_via_va_list` functions: https://github.com/zephyrproject-rtos/zephyr/blob/fdc25cd44c2b693aecff96d3dfcd0ee088d48d5b/lib/os/cbprintf_packaged.c#L75

Given that maybe instead of `cbprintf_locate/settle` approach, it would be better to have something like `cbpprintf_external(ext_formatter func, void *ctx, void *packaged)` where external formatter would be function like `foo(void *ctx, const char *fmt, va_list ap)`.

dcpleung replied:
I have updated `cbprintf_packaged.c` to use external formatter. Hopefully this aligns with what you are suggesting.

dcpleung committed:
lib/os: cbprintf_packaged: add ability to use external formatter

This adds to the cbprintf_packaged library to allow external
formatters to be used by the way of callback. This will allow
logging backends to use their own formatter for output if
needed.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
logging: syst: fix logging v2 with strings as arguments

The packaged string coming from logging subsys requires some
additional processing if there are string arguments.
These strings are actually embedded inside the package so
the string pointers inside the argument list must be
replaced with pointers to strings inside the package.
Without this extra step of processing, MIPI Sys-T's printf
function would process these arguments directly which may be
NULL pointers or invalid one pointing to somewhere.

This utilizes the new cbpprintf_external() for the processing
before feeding data to the MIPI library.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

npitre reviewed (approved):
OK but please fix the typo in .

nordic-krch reviewed (approved):
looks good.
